### ISSUE TITLE
Universal Head Service

### ISSUE THREAD

**Issue by jeffwhelpley** [angular/angular#7301]
We need a universal service (i.e. both on client and server) that can be used by any Angular 2 app to manage the HEAD section of the HTML document. For example:

```
class MyHomePage {
  constructor(headService: HeadService) {
    headService.setTitle('My home page title');
    headService.setDescription('Description here');
    headService.setMeta('author', 'Jeff Whelpley');
  }
}
```

You should also be able to set these values declaratively in the component decorator or potentially as part of the router. This issue is for coming up with a design for this service and implementing it in core. Once that is in place, the server side implementation of HeadService will be implemented on the Angular Universal repo:

https://github.com/angular/universal/issues/238

**Comment by jeffwhelpley** [comment]
I talked with @jeffbcross and @gdi2290 on Friday about this as well as #7302 and #7303. At a high level, before talking about design for these services, we need to figure out at an even higher level where we want the tokens and interfaces/abstract implementations for these types of things to live. The actual implementations of the head service or any other can live outside core and can be easily overridable by anyone, but I think the key is ensuring a common interface is available for everyone.

**Comment by pkozlowski-opensource** [comment]
Related: we've got already a service that handle title read / write: https://github.com/angular/angular/blob/4a93f58b8bce4e872c8280c4971c989b37e4936d/modules/angular2/src/platform/browser/title.ts

**Comment by jeffwhelpley** [comment]
Good to at least have the title service which is obviously the primary piece of the head that matters on the client side but we ideally can at least expand the interface to include everything in the head even if what is included in the core implementation only has the title.

**Comment by manekinekko** [comment]
Great. 

A. For the imperative approach, I think we could easily expand the current Title service and have a Head service with a richer API as required above. This update would involve touching the following files (I guess):

1. modules/angular2/src/platform/browser/title.ts
1. modules/angular2/src/platform/browser/browser_adapter.ts
1. modules/angular2/src/platform/dom/dom_adapter.ts 
1. modules/angular2/src/platform/server/parse5_adapter.ts 
1. modules/angular2/test/platform/browser/title_spec.ts

B. For the declarative approach, this would need more work I guess as we need to update the core files : [view_resolver.ts](modules/angular2/src/core/linker/view_resolver.ts) and [metadata/view.ts](modules/angular2/src/core/metadata/view.ts) (at least).

@jeffwhelpley what do you think?

**Comment by intellix** [comment]
Would be good to have an array of a particular meta tag, we're doing this for handling og:images: 
`<meta property="og:image" content="{{image}}" ng-repeat="image in head.meta.images">`

**Comment by radek-holy** [comment]
FTR, the request for manipulating <meta> tags was closed as "not feasible": #3144

**Comment by radek-holy** [comment]
Would it be possible extend the service over the <html> tag as well, please?

Two use cases follow:

1. I'm using a CSS framework which requires different classes of the <html> element for different layouts. So, when Angular's router navigates to different components with different layouts, I need to change the classes of the root element.
2. An i18n service would probably like to change the `lang` attribute of the root element.

(please, let me know if I should rather file another issue)

**Comment by jeffwhelpley** [comment]
Closing in favor of #7438. Btw, @radek-holy regarding control over the html element, @gdi2290 and I are experimenting with the idea of a top level "root" app. We will open up other issues depending on what we find.

**Comment by radek-holy** [comment]
Cool. If you remember, please, let me know when you open such an issue.

**Comment by jeffwhelpley** [comment]
@radek-holy there are going to be a number of smaller issues and PRs opened for this effort. Some things we already started working on:

https://github.com/angular/angular/pull/7455
https://github.com/angular/universal/issues/309
https://github.com/angular/universal/issues/308

Basically, we are taking the approach of getting this working on the Universal repo side and then whenever we encounter a specific roadblock in core, we open an issue and/or PR just for that smaller thing.

**Comment by fulls1z3** [comment]
I have developed a meta utility ([@nglibs/meta](https://github.com/nglibs/meta)) in last November, which is generating the meta tags based on route definitions.

Just updated it last weekend, to avoid direct manipulation of DOM. It now depends on DomAdapter in Angular 2 and on Meta/Title services in Angular 4. And works well on server-side rendering as well as client-side rendering.

Feel free to have a look and use it, probably it will provide required functionality.
